source/common/http/filter/ip_tagging_filter.cc

```diff
@@ -9,7 +11,38 @@ IpTaggingFilter::~IpTaggingFilter() {}
 
 void IpTaggingFilter::onDestroy() {}
 
-FilterHeadersStatus IpTaggingFilter::decodeHeaders(HeaderMap&, bool) {
+FilterHeadersStatus IpTaggingFilter::decodeHeaders(HeaderMap& headers, bool) {
+  bool is_internal_request =
+      headers.EnvoyInternalRequest() && (headers.EnvoyInternalRequest()->value() == "true");
+
+  if ((is_internal_request && config_->requestType() == FilterRequestType::EXTERNAL) ||
+      (!is_internal_request && config_->requestType() == FilterRequestType::INTERNAL) ||
+      (!config_->runtime().snapshot().featureEnabled("ip_tagging.http_filter_enabled", 100))) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  std::vector<std::string> tags =
+      config_->trie().getTags(callbacks_->requestInfo().downstreamRemoteAddress());
+
+  if (!tags.empty()) {
+    HeaderString& header = headers.insertEnvoyIpTags().value();
+    if (!header.empty()) {
+      header.append(", ", 2);
+    }
+    if (tags.size() > 1) {
+      std::string tags_join = absl::StrJoin(tags, ", ");
+      header.append(tags_join.c_str(), tags_join.size());
+    } else {
+      header.append(tags[0].c_str(), tags[0].size());
+    }
+
+    for (const std::string& tag : tags) {
+      config_->scope().counter(fmt::format("{}{}.hit", config_->statsPrefix(), tag)).inc();
```

Comment:
i worry about stat cardinality if someone ends up with thousands of tags. can you add a comment if we're going to keep it this way?

Reply:
I'll leave it as is to provide  visibility into what is being tagged. Once ip tags are logged in the grpc access logs that can be the other way to track what is being tagged.  Then the configuration or runtime key or some other way can expose should this visibility be exposed via stats, logging or both.